### PR TITLE
Add ISO regions (us-iso-east-1, us-isob-east-1)

### PR DESCRIPTION
**Pull request by Callisto13**
### TLDR 
- Add isolated regions and partition mappings
- Fetch account id from environment if isolated region is set (`ISO_EKS_ACCOUNT_ID`, name up for debate)
  - Try to fail fast when this is not set

### Detail
This PR adds the ISO regions and partitions, and also includes a change to the way we fetch [Resource/Account/Owner IDs](https://github.com/weaveworks/eksctl/blob/master/pkg/apis/eksctl.io/v1alpha5/types.go#L220-L244). This is because the account numbers for the AMIs which can be used in these secret isolated regions are also secret and cannot be made public.

All the other account IDs (for this particular AMI family) are simply stored in constants where they can be called from anywhere in the code... and they are, from just about everywhere. So adding a once-in-a-while user input flow is a little fiddly. I spiked on/considered several approaches:

1. The ["F the UX" plan](https://github.com/Callisto13/eksctl/commit/8ea941d2efb91c1049ab7244089677ec4138bcd2)
    - This involves checking the env in the same place we would normally just return a constant. It sucks because, if you had forgotten to set the env var, you would not know about it until _after_ your stack had been created.
1. The ["Passing it through everywhere" plan](https://github.com/Callisto13/eksctl/commit/8d312581d7ee78bec2c826296232226a0334a382)
    - This involves checking the region and validating the env var (or not) at the "top" of any command which would end up needing it (there are 5) and passing it down through the various threads. This is technically the "correct" thing to do, but since this account is used _everywhere_ and quite a long way down sometimes, it is super invasive. In one instance it may end up not being used at all, depending on whether a different image family was chosen for the nodegroup.
      - Since you ask, setting the "true" account id (based on both region _and_ AMI family) is not possible since each node group could ask for a different family, therefore there is no "true" id which can be applied everywhere.
1. The ["Java-style getter/setter" plan](https://github.com/Callisto13/eksctl/commit/da452579894e6cffe155635e472d488cb0aca78a)
    - In which the account ID is set once at the very beginning, stored in a private variable which can then be called via a public function when needed. This feels like an odd pattern in Go.
1. The "Sneaky init" plan (no spike)
    - Similar to the above, but done in an `init()` at the start. This would only be possible if I moved the whole flag/config loading into an init too, because I would need the region to decide whether or not to find the env var. I decided moving our already complicated cmd starting points into an init would be a deathtrap as inits are very hard to debug when something goes wrong.
1. **The "Just check twice" plan (this PR)**
    - Similar to plan 1, except we also check for the presence of the env var at the start of the call if an iso region is set.

I chose the approach in this PR (5) based on the fact I hated it the least. I am happy to change things if anyone has a better idea.

## Questions
1. Any thoughts on the name `ISO_EKS_ACCOUNT_ID` for the env var?
1. How do we want to document this? _Do_ we want to document it? Where?
1. Also: have I missed anything for adding the regions?

### Checklist
- [x] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [ ] Manually tested (has to be sent to aws)
- [x] Added labels for change area (e.g. `area/nodegroup`), target version (e.g. `version/0.12.0`) and kind (e.g. `kind/improvement`)
- [x] Make sure the title of the PR is a good description that can go into the release notes

### CODE REVIEW

**Comment by Callisto13** [comment]
Any thoughts on this?

**Comment by cmdallas** [comment]
1. Any thoughts on the name ISO_EKS_ACCOUNT_ID for the env var?
> +1. Currently, all 'special' aws regions (existing & roadmap) use the an 'iso' convention in their arn partitions.

2. How do we want to document this? Do we want to document it? Where?
> AWS has internal docs for customers of special regions that generally document these kinds of differences. This is a bit of a unique scenario, but I think it makes sense for us (AWS) to add this to our diff docs. I will take an action item to get this published in the internal docs.

**Comment by Callisto13** [comment]
👍  thanks @cmdallas sounds good

**Comment by Callisto13** [comment]
**Issue:** `eksctl get clusters --all-regions` errors (see below) for regular folks because they cannot run inside the ISO networks. The SDK's send request fails repeatedly but [will retry](https://github.com/aws/aws-sdk-go/blob/master/aws/corehandlers/handlers.go#L162-L164) because that is the standard behaviour for network errors. (We do not see this retry behaviour for Gov or CN regions; we get the expected `403` back in those cases.)

```
```

**Options:**
a) Change to EKSCTL: We exclude the ISO regions when calling `get clusters --all-regions` (while we are at it: the UX for getting Gov and CN regions when you do not have creds is a bit rubbish)
b) Change to aws-cdk: We set ISO-specific retry rules.

All thoughts and votes and new ideas welcome please!

**Comment by mikestef9** [comment]
I prefer option a, would also like to get @cmdallas thoughts as well

**Comment by cPu1** [comment]
Another option is to use `EC2:DescribeRegions` to get all the regions enabled for the account, and run tests on those regions only. Although it's unclear whether it would return isolated regions as well if an account has access to them.

**Comment by Callisto13** [comment]
Ooh thanks @cPu1 ! That would solve all UX annoyances 😄

File: pkg/apis/eksctl.io/v1alpha5/account.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	return nil
+}
+
+func lookupIsoAccountID(region string) (string, error) {
+	id, ok := os.LookupEnv(IsoEKSAccountIDEnv)
+	if !ok {
+		return "", errors.Errorf("%s not set, required for use of region: %s", IsoEKSAccountIDEnv, region)
+	}
```
Comment (cPu1) [inline pkg/apis/eksctl.io/v1alpha5/account.go:70]: This will pass validation if the env var is set but is empty (e.g., `ISO_EKS_ACCOUNT_ID= eksctl create cluster`). I think we should validate that the value is non-empty.

**Review by cPu1 — APPROVED** [review]

File: pkg/apis/eksctl.io/v1alpha5/account.go (hunk 1 above)
Comment (Callisto13) [reply pkg/apis/eksctl.io/v1alpha5/account.go:70]: 👍  thanks

**Comment by cmdallas** [comment]
@Callisto13 

Failing in region because a us-west-2 sts endpoint is being called:

```
export ISO_EKS_ACCOUNT_ID=<redacted>
./eksctl get cluster --name=<redacted>
Error: checking AWS STS access - cannot get role ARN for current session: RequestError: send request failed
cause by Post "https://sts.us-west-2.amazonaws.com"
```

https://github.com/weaveworks/eksctl/blob/d4b46b2c84a6175cd54d3038d8074413bdadfddc/pkg/eks/api.go#L264

**Comment by Callisto13** [comment]
Thanks @cmdallas I will look into it and get a new build back to you asap 👍

**Comment by Callisto13** [comment]
@cmdallas I think the region defaults to `us-west-2`, could you see what happens when you pass in the `--region` flag?

**Comment by cmdallas** [comment]
My bad!

Cool, `get` commands are working fine. I attempted a `create` command, but we are missing SSM params in region. I will try and get that updated by Monday the 13th

**Comment by cmdallas** [comment]
Sorry to slip!! I am out on vacation and will not be back for a few weeks. I will pass this on to my team for the time being.

**Comment by Callisto13** [comment]
No worries @cmdallas, pretty much everyone on our side is also on vacation next week. I'll be back on the 29th.

**Comment by Callisto13** [comment]
Gonna close this an re-implement when upstream APIs support these regions.
